lib/streamlit/state/session_state.py

```diff
@@ -14,25 +14,169 @@
 
 from typing import (
     Any,
+    ItemsView,
+    KeysView,
+    ValuesView,
     cast,
     Dict,
     Iterator,
     MutableMapping,
     Optional,
     TYPE_CHECKING,
     Union,
+    Tuple,
+    Callable,
+    Set,
+    List,
 )
 
 import attr
 
 from streamlit.errors import StreamlitAPIException
-from streamlit.report_thread import get_report_ctx, ReportContext
+from streamlit.proto.WidgetStates_pb2 import WidgetState as WidgetStateProto
+from streamlit.proto.WidgetStates_pb2 import WidgetStates as WidgetStatesProto
 
 if TYPE_CHECKING:
     from streamlit.report_session import ReportSession
 
 
-@attr.s(auto_attribs=True)
+@attr.s(auto_attribs=True, slots=True, frozen=True)
+class Serialized:
+    value: WidgetStateProto
+
+
+@attr.s(auto_attribs=True, slots=True, frozen=True)
+class Value:
+    value: Any
+
+
+WState = Union[Serialized, Value]
+
+WidgetArgs = Tuple[Any, ...]
+WidgetCallback = Callable[..., None]
+WidgetDeserializer = Callable[[Any], Any]
+WidgetSerializer = Callable[[Any], WidgetStateProto]
+WidgetKwargs = Dict[str, Any]
+
+
+@attr.s(auto_attribs=True, slots=True, frozen=True)
+class WidgetMetadata:
+    id: str
+    deserializer: WidgetDeserializer
+    serializer: WidgetSerializer
+    value_type: Any
+    has_key: bool = False
+
+    callback: Optional[WidgetCallback] = None
+    callback_args: Optional[WidgetArgs] = None
+    callback_kwargs: Optional[WidgetKwargs] = None
+
+
+@attr.s(auto_attribs=True, slots=True)
+class WStates(MutableMapping[str, Any]):
+    states: Dict[str, WState] = attr.Factory(dict)
+    widget_metadata: Dict[str, WidgetMetadata] = attr.Factory(dict)
+
+    def __getitem__(self, k: str) -> Any:
+        item = self.states.get(k, None)
+        if item is not None:
+            if isinstance(item, Value):
+                return item.value
+            else:
+                metadata = self.widget_metadata.get(k, None)
+                if metadata is None:
+                    # No deserializer, which should only happen if state is gotten from a reconnecting browser
+                    # and the script is trying to access it. Pretend it doesn't exist.
+                    raise KeyError
+                deserialized = metadata.deserializer(
+                    item.value.__getattribute__(metadata.value_type)
+                )
+                self.states[k] = Value(deserialized)
+                return deserialized
+        else:
+            raise KeyError
+
+    def __setitem__(self, k: str, v: WState):
+        self.states[k] = v
+
+    def __delitem__(self, k: str) -> None:
+        del self.states[k]
+
+    def __len__(self) -> int:
+        return len(self.states)
+
+    def __iter__(self):
+        for key in self.states:
+            yield key
+
+    def keys(self) -> Set[str]:
+        return set(self.states.keys())
+
+    def items(self) -> Set[Tuple[str, Any]]:
+        i = [(k, self[k]) for k in self]
+        return set(i)
+
+    def values(self) -> Set[Any]:  # type: ignore
```

Comment:
Hm, I'm surprised to see that a `# type: ignore` is needed here

Reply:
Mypy expects a `ValuesView` because of the `MutableMapping` abc, and doesn't consider `Set` a valid subset. Don't know why the other two are ok though.